### PR TITLE
feat: reusable workflows (ADR 31), centralized routing (ADR 34), layered content (ADR 35)

### PR DESCRIPTION
**Pull request by waynesun09**
Implementation of ADR 31 (reusable workflows), ADR 34 (centralized routing), and ADR 35 (layered content resolution) for the fullsend agent pipeline.

### ADR 31 — Reusable workflows for action-installed distribution

- Publish 5 reusable workflows (`workflow_call`) and a root composite action from `fullsend-ai/fullsend`, enabling org `.fullsend` repos to delegate agent pipeline logic upstream instead of maintaining full scaffold copies
- Convert scaffold agent workflows from 80-289 line full implementations to ~30-55 line thin callers that use `workflow_call` + explicit `secrets:` passthrough
- Extract 3 shared composite actions (`mint-token`, `validate-enrollment`, `setup-gcp`) to `.github/actions/` for use by reusable workflows
- Consolidate duplicate fullsend action to single root `action.yml` (was duplicated at `.github/actions/fullsend/` and scaffold copy)
- Remove CLI version-pinning from `WorkflowsLayer` since thin callers no longer contain CLI download steps — version is now controlled by `fullsend_version` input to reusable workflows

### ADR 34 — Centralized event routing via dispatch.yml

- Move event-to-stage routing from the per-target-repo shim (8 jobs, ~210 lines) into `dispatch.yml` in the `.fullsend` repo
- Shim shrinks to 2 jobs (~80 lines): a universal `dispatch` forwarder via `workflow_call` and a `stop-fix` handler
- New "Determine stage" step in dispatch.yml routes events to stages using shell `case` logic with `env:` blocks (no expression injection)
- Adding a new stage requires only a case branch in dispatch.yml — zero changes to enrolled repos
- No backwards compatibility needed: shim and dispatch.yml ship together atomically via `fullsend admin install`

### ADR 35 — Layered content resolution

- Upstream defaults (agents, skills, schemas, harness, policies, scripts, env) are no longer copied into `.fullsend` repos by the scaffold
- Reusable workflows sparse-checkout upstream defaults from `fullsend-ai/fullsend@v0` at runtime and layer org overrides from `customized/` on top
- Composite actions are referenced directly from upstream (`fullsend-ai/fullsend/.github/actions/*@v0`), not installed into `.fullsend`
- Scaffold output shrinks from ~68 files to ~23 org-specific files (`customized/` gitkeeps, thin callers, config, env files)

### New files (repo root, published upstream)

| File | Description |
|---|---|
| `action.yml` | Published composite action (install CLI, run agent, upload artifacts) — single source, replaces duplicate at `.github/actions/fullsend/` |
| `.github/actions/mint-token/action.yml` | OIDC token mint via fullsend mint service |
| `.github/actions/validate-enrollment/action.yml` | Config.yaml enrollment validation |
| `.github/actions/setup-gcp/action.yml` | GCP WIF auth + credential masking |
| `.github/workflows/reusable-{triage,code,review,fix,retro}.yml` | Reusable agent pipelines |
| `docs/ADRs/0035-layered-content-resolution.md` | ADR 35 decision record |

### Modified files

| File | Change |
|---|---|
| `internal/scaffold/.../workflows/{triage,code,review,fix,retro}.yml` | Reduced to thin callers (~30-55 lines) |
| `internal/scaffold/.../workflows/dispatch.yml` | Added centralized routing logic with `event_action` input |
| `internal/scaffold/.../workflows/{prioritize,prioritize-scheduler,repo-maintenance}.yml` | Updated to use upstream action refs (`fullsend-ai/fullsend/.github/actions/*@v0`) and workspace layering for scripts |
| `internal/scaffold/.../templates/shim-workflow-call.yaml` | Simplified from 8 jobs to 2 jobs (~80 lines) |
| `internal/scaffold/.../customized/{agents,skills,schemas,harness,policies,scripts,env}/.gitkeep` | Override directories for org customization |
| `internal/scaffold/scaffold.go`, `scaffold_test.go` | Updated tests for thin callers, routing logic, shim structure, slim scaffold output |
| `internal/layers/workflows.go` | Removed version-pinning logic |
| `internal/cli/admin.go` | Updated `NewWorkflowsLayer` call sites |
| `internal/mint/main.go`, `main_test.go` | Accept `fullsend-ai/fullsend/` prefix in `job_workflow_ref` for cross-org `workflow_call` |

### Deleted files

| File | Reason |
|---|---|
| `.github/actions/fullsend/action.yml` | Consolidated to root `action.yml` |
| `internal/scaffold/fullsend-repo/.github/actions/fullsend/action.yml` | Scaffold copy of above, also consolidated |

### Architecture

```
target-repo shim (2 jobs, ~80 lines)
  └─ workflow_call ──> .fullsend/dispatch.yml
                          │  "Determine stage" routing
                          │  OIDC mint + config checks
                          │
          ┌───────────────┼───────────────────────┐
          ▼               ▼                       ▼
  .fullsend/code.yml   review.yml          triage.yml
  (~30-55 lines, thin callers via workflow_call)
          │
  uses: fullsend-ai/fullsend/.github/workflows/reusable-code.yml@v0
          │                    ┌─ sparse-checkout upstream defaults (ADR 35)
          │                    ├─ layer customized/ overrides on top
          │                    └─ clean up .defaults directory
          │
  uses: fullsend-ai/fullsend@v0  ← published root composite action
```

**ADR 31 properties:**
- `vars.*` mapped to explicit `workflow_call` inputs (vars don't cross the boundary)
- Secrets pass via explicit `secrets:` blocks (not `secrets: inherit` — doesn't work cross-org)
- 1 level of `workflow_call` nesting (limit is 4)
- OIDC `job_workflow_ref` reports the reusable workflow's repo — mint accepts both `{org}/.fullsend/` and `fullsend-ai/fullsend/` prefixes

**ADR 34 routing:**
- Shim forwards `event_action` (not `stage`) to dispatch.yml
- Routing uses `github.event.*` context via `env:` blocks — no expression injection
- Kill switch and role gating in dispatch.yml
- Fork PR detection for fix stage
- Author association checks for `/fix` and `/retro` commands
- Stage-to-role mapping: `retro` and `prioritize` map to `fullsend` App role (no dedicated PEM)

**ADR 35 layering:**
- Reusable workflows sparse-checkout `fullsend-ai/fullsend@v0` at runtime for upstream defaults
- Org overrides in `customized/` are layered on top
- Non-agent workflows (repo-maintenance, prioritize) use workspace layering for scripts
- Composite actions referenced directly from upstream, not installed into `.fullsend`

## Test plan

- [x] `go test ./internal/...` — all 19 packages pass
- [x] `go vet ./internal/...` — clean
- [x] Review squads (security, correctness, test coverage, YAML consistency, parity) — multiple rounds, clean
- [x] Manual: `fullsend admin install` against nonflux org — 5 apps reused, 25 scaffold files written (down from 77), mint health check passed, enrollment completed (2 repos)
- [x] Scaffold verification: thin callers are 41-66 lines with 1 reusable ref each; dispatch.yml has routing step; `customized/` override directories all present with `.gitkeep`
- [x] Mint redeployment: `--force-mint-deploy` required when upgrading from PR #503 — the mint's `job_workflow_ref` validation must accept `fullsend-ai/fullsend/` prefix for cross-org `workflow_call`. Without redeployment, reusable workflows get 403 from the old mint.
- [x] End-to-end dispatch chain (triage on nonflux/integration-service): shim → dispatch.yml routing → triage.yml thin caller → `reusable-triage.yml@v0` → sparse-checkout upstream defaults → mint-token action → GCP auth → agent setup — all green. Agent run fails due to openshell gateway incompatibility (#780), unrelated to this PR.
- [x] Rebase on main after direct-WIF merge (PR #856) — resolved conflicts, aligned reusable workflows and setup-gcp with WIF-only auth (removed `gcp_auth_mode`, `FULLSEND_GCP_WIF_SA_EMAIL`, `FULLSEND_GCP_SA_KEY_JSON`)
- [ ] Dogfood: replace `fullsend-ai/.fullsend` workflows with thin callers, trigger dispatch chain
- [ ] Test remaining stages end-to-end: code, review, fix, retro, stop-fix
- [ ] Verify routing negative cases: bot filtering, /code on PR, unauthorized /fix, fork PR, kill switch

### Issues found and fixed during testing

1. **`/health` endpoint (GCF routing)** — CLI polled `/healthz` which GCF Functions Framework v1.9.0 doesn't route to the handler. Fixed: changed to `/health` in `gcf.go`.
2. **Missing permissions blocks in thin callers** — org default workflow permissions are `read`; `workflow_call` caps the called workflow's permissions to the caller's top-level `permissions:` block. Without explicit permissions, reusable workflows got `none` for write scopes. Fixed: added `permissions:` blocks to all thin caller templates.
3. **`secrets: inherit` doesn't work cross-org** — `nonflux/.fullsend` calling `fullsend-ai/fullsend` is cross-org; `secrets: inherit` only works within the same org/enterprise. Fixed: reusable workflows declare explicit `secrets:` inputs; thin callers pass each secret individually.
4. **`uses: ./` resolves to caller's checkout** — in `workflow_call`, `uses: ./` resolves to the workspace root (caller's `.fullsend` checkout), not the reusable workflow's repo. Fixed with upstream action refs (`fullsend-ai/fullsend/.github/actions/*@v0`).
5. **Cross-org `job_workflow_ref` OIDC validation** — mint validated `job_workflow_ref` starts with `{org}/.fullsend/`, but cross-repo `workflow_call` reports `fullsend-ai/fullsend/.github/workflows/reusable-*.yml@...`. Fixed: updated mint to also accept `fullsend-ai/fullsend/` prefix.
6. **App manifest permissions drift** — `nonflux-coder`/`nonflux-review` Apps created with `issues: read`, but mint requests `issues: write`. Fixed: upgraded manifests in `types.go` (33920c8e). Existing Apps need manual permission update.
7. **`fullsend` composite action not at repo root** — reusable workflows reference `fullsend-ai/fullsend@v0`, consolidated the action to root `action.yml` (removed duplicates at `.github/actions/fullsend/` and scaffold copy).
8. **Non-agent workflows used local action refs** — `repo-maintenance.yml`, `prioritize.yml`, `prioritize-scheduler.yml` used `./.github/actions/mint-token` (local ref), but ADR 35 no longer installs composite actions into `.fullsend`. Fixed: updated to upstream action refs and added workspace layering for scripts.
9. **retro/prioritize stages blocked by dispatch role check** — `dispatch.yml` maps stage→role but only handled `code → coder`. `retro` and `prioritize` share the `fullsend` App. Fixed: added `retro|prioritize → fullsend` mapping and updated `reusable-retro.yml` to mint with `role: fullsend`.
10. **Mint 403 after `--skip-mint-deploy`** — upgrading from PR #503 to PR #792 with `--skip-mint-deploy` leaves the old mint code deployed, which rejects `fullsend-ai/fullsend/` in `job_workflow_ref`. Must use `--force-mint-deploy` on first install after this PR.
11. **Stale direct-WIF references after rebase** — after rebasing on main (which merged PR #856 direct-WIF), reusable workflows and setup-gcp still had `gcp_auth_mode`, `FULLSEND_GCP_WIF_SA_EMAIL`, `FULLSEND_GCP_SA_KEY_JSON`. Fixed: aligned all reusable workflows and setup-gcp with WIF-only auth.

### CODE REVIEW

**Comment by waynesun09** [comment]
## End-to-end test results — nonflux org (2026-05-09, updated 2026-05-10)

Tested the full reusable workflow + centralized routing + layered content resolution (ADRs 30, 33, 34) against the `nonflux` org. Binary built from this branch, installed via `fullsend admin install`, all `@v1` refs manually updated to `@adr-0030-reusable-workflows` for pre-merge testing.

### Install

- 25 scaffold files written (down from 77 pre-ADR-34)
- No agents, skills, schemas, harness, policies, or scripts installed (ADR 34 — provided at runtime)
- `customized/` directory with `.gitkeep` files for org overrides
- Composite actions NOT in `.fullsend` — referenced directly from upstream
- Mint health check passed, repo enrollment completed

### Dispatch chain results

| Stage | Shim | Routing | Thin caller | Upstream defaults | Mint token | GCP auth | Agent setup | Agent run |
|-------|:----:|:-------:|:-----------:|:-----------------:|:----------:|:--------:|:-----------:|:---------:|
| `/triage` | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: #780 |
| `/code` | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: #780 |
| `/review` | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: #780 |
| `/fix` | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: #780 |
| `/retro` | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: #780 |
| `/stop-fix` | :white_check_mark: | :white_check_mark: | N/A | N/A | N/A | N/A | N/A | :white_check_mark: |
| `repo-maintenance` | N/A | N/A | N/A | :white_check_mark: (scripts) | :white_check_mark: | N/A | N/A | :white_check_mark: |

All 5 agent-stage failures are openshell gateway incompatibility (#780), not related to the reusable workflow changes. The full dispatch chain (shim → dispatch.yml routing → thin caller → reusable workflow → sparse-checkout upstream defaults → workspace layering → upstream composite actions → agent run) works end-to-end.

### Issues found and fixed (9 total)

1. **`/health` endpoint (GCF routing)** — CLI polled `/healthz` but GCF Functions Framework v1.9.0 doesn't route it. Fixed: changed to `/health`.

2. **Missing permissions blocks in thin callers** — org defaults are `read`; `workflow_call` caps callee permissions to caller's top-level block. Fixed: added `permissions:` to all thin callers.

3. **`secrets: inherit` doesn't work cross-org** — cross-org `workflow_call` requires explicit `secrets:` passthrough. Fixed: reusable workflows declare explicit `secrets:` inputs.

4. **`uses: ./` resolves to caller's checkout** — in `workflow_call`, `./` is the caller's workspace, not the reusable workflow's repo. Superseded by ADR 34 which references actions directly from upstream.

5. **Cross-org `job_workflow_ref` OIDC validation** — mint only accepted `{org}/.fullsend/` prefix. Fixed: also accept `fullsend-ai/fullsend/` prefix.

6. **App manifest permissions drift** — `issues: read` in App manifests vs `issues: write` in rolePermissions. Fixed: upgraded to `write` in `types.go`.

7. **`fullsend` composite action not at repo root** — action only existed in scaffold, not at `.github/actions/fullsend/`. Fixed: copied to repo root.

8. **Non-agent workflows used local action refs (ADR 34)** — `repo-maintenance.yml`, `prioritize.yml`, `prioritize-scheduler.yml` still used `./.github/actions/mint-token` but ADR 34 no longer installs actions. Fixed: updated to upstream refs + workspace layering for scripts.

9. **retro/prioritize stages blocked by dispatch role check** — dispatch.yml only mapped `code → coder` but `retro` and `prioritize` share the `fullsend` App (no dedicated Apps). Fixed: added `retro|prioritize → fullsend` mapping. Also updated `reusable-retro.yml` to mint with `role: fullsend`.

**Comment by waynesun09** [comment]
### Install verification

`fullsend admin install` (v0.7.0-102-gf3f50464) against `nonflux` org completed successfully:

- Token permissions verified
- All 5 Apps found and reused
- **25 scaffold files written** (slim scaffold — ADR 34):
  - Thin callers (triage, code, review, fix, retro)
  - `dispatch.yml` with centralized routing
  - `repo-maintenance.yml`, `prioritize.yml`, `prioritize-scheduler.yml` with upstream action refs
  - `customized/` gitkeeps for org overrides (agents, skills, schemas, harness, policies, scripts)
  - `env/` files, `config.yaml`, `AGENTS.md`, `CODEOWNERS`, shim template
  - **No** agents, skills, schemas, harness, policies, scripts (provided at runtime)
  - **No** `.github/actions/` (referenced directly from upstream)
- Mint health check passed on first attempt
- Org variable `FULLSEND_MINT_URL` stored
- Repo enrollment completed (2 repos) via `repo-maintenance.yml` with upstream action refs + workspace layering
- Permission check correctly detected App manifest drift and blocked until updated

### Workflow runs

- Repo Maintenance (enrollment): https://github.com/nonflux/.fullsend/actions/runs/25609252404
- Triage: https://github.com/nonflux/.fullsend/actions/runs/25609265599
- Code: https://github.com/nonflux/.fullsend/actions/runs/25609267628
- Review: https://github.com/nonflux/.fullsend/actions/runs/25609265885
- Fix: https://github.com/nonflux/.fullsend/actions/runs/25609266244

**Comment by waynesun09** [comment]
## Integration test results — nonflux org (2026-05-09, updated 2026-05-10)

Tested `fullsend admin install` (v0.7.0-dev) against the `nonflux` org with full end-to-end dispatch chain validation.

### Install

- All 5 Apps found and reused
- 25 scaffold files written (thin callers, dispatch.yml, shim template, customized/ gitkeeps, env files, config.yaml) — down from 77 pre-ADR-34
- Mint health check passed
- Permission check correctly detected `nonflux-coder` and `nonflux-review` missing `issues:write`, blocked install until updated

### End-to-end dispatch chain

| Stage | Shim | Routing | Thin caller | Upstream defaults | Mint token | GCP auth | Agent setup | Agent run |
|-------|:----:|:-------:|:-----------:|:-----------------:|:----------:|:--------:|:-----------:|:---------:|
| `/triage` | pass | pass | pass | pass | pass | pass | pass | fail (#780) |
| `/code` | pass | pass | pass | pass | pass | pass | pass | fail (#780) |
| `/review` | pass | pass | pass | pass | pass | pass | pass | fail (#780) |
| `/fix` | pass | pass | pass | pass | pass | pass | pass | fail (#780) |
| `/retro` | pass | pass | pass | pass | pass | pass | pass | fail (#780) |
| `/stop-fix` | pass | pass | N/A | N/A | N/A | N/A | N/A | pass |
| repo-maintenance | N/A | N/A | N/A | pass (scripts) | pass | N/A | N/A | pass |

All agent-run failures are the openshell gateway incompatibility (#780), unrelated to this PR.

### Issues found and fixed (9 total)

1. **`/health` endpoint** — GCF Functions Framework v1.9.0 doesn't route `/healthz`. Fixed: use `/health`.
2. **Missing permissions blocks in thin callers** — `workflow_call` caps permissions to caller's top-level block. Fixed: added explicit `permissions:` to all thin caller templates.
3. **`secrets: inherit` cross-org** — doesn't work across orgs. Fixed: explicit `secrets:` inputs in reusable workflows, thin callers pass individually.
4. **`uses: ./` resolves to caller** — in `workflow_call`, local action refs resolve to the caller's checkout. Fixed: reference actions directly from upstream (`fullsend-ai/fullsend/.github/actions/*@v1`).
5. **Cross-org `job_workflow_ref` OIDC** — OIDC claim reports the reusable workflow's repo. Fixed: mint accepts both `{org}/.fullsend/` and `fullsend-ai/fullsend/` prefixes.
6. **App manifest permissions drift** — `types.go` had `issues: read` vs `main.go` `issues: write`. Fixed: upgraded manifests to `issues: write` (33920c8e).
7. **Composite action not at repo root** — reusable workflows reference `.github/actions/fullsend@v1` but it only existed in scaffold. Fixed: copied to repo root (fc27a5ee / 7f3fdd9b).
8. **Non-agent workflows used local action refs** — `repo-maintenance.yml`, `prioritize.yml`, `prioritize-scheduler.yml` used `./.github/actions/mint-token` but ADR 34 no longer installs actions. Fixed: upstream refs + workspace layering for scripts (f0d111e2).
9. **retro/prioritize blocked by dispatch role check** — `dispatch.yml` only mapped `code → coder`. Fixed: added `retro|prioritize → fullsend` mapping + reusable-retro mints with `role: fullsend` (cde89b90).

### Dispatch chain verified

```
target-repo shim (2 jobs)
  → workflow_call → .fullsend/dispatch.yml (routing + fan-out)
    → workflow_dispatch → .fullsend/triage.yml (thin caller)
      → workflow_call → fullsend-ai/fullsend/reusable-triage.yml (pipeline logic)
        → sparse-checkout upstream defaults + layer customized/ overrides
        → uses: fullsend-ai/fullsend/.github/actions/fullsend@v1 (run agent)
```

**Comment by waynesun09** [comment]
@ralphbean @ifireball the PR based on the #503 token mint for reusable workflow, test pass and coverage in last comment.

File: internal/mint/main.go [hunk 1]
```diff
@@ -565,20 +565,27 @@ func (h *Handler) prevalidateOIDCToken(token string) (*oidcClaims, error) {
...
 	ref := strings.ToLower(claims.JobWorkflowRef)
-	expectedPrefix := strings.ToLower(claims.RepositoryOwner) + "/.fullsend/"
-	if !strings.HasPrefix(ref, expectedPrefix) {
-		return nil, fmt.Errorf("job_workflow_ref does not reference .fullsend repo")
-	}
-
-	// Extract the workflow file path relative to .fullsend/ (before @ref).
-	relPath := strings.TrimPrefix(ref, expectedPrefix)
+	configPrefix := strings.ToLower(claims.RepositoryOwner) + "/.fullsend/"
+	upstreamPrefix := "fullsend-ai/fullsend/"
```
Comment (ralphbean) [inline internal/mint/main.go:577]: **[important]** This PR correctly extends `prevalidateOIDCToken` to accept upstream `fullsend-ai/fullsend` workflow refs (necessary for reusable workflows via `workflow_call`). However, the `ALLOWED_WORKFLOW_FILES` check downstream (line 597) still fails open — when unset, `h.allowedWorkflows` is empty and the entire allowlist block is skipped, so **any** workflow file under `.github/workflows/` in `fullsend-ai/fullsend` can mint agent tokens for any org in `ALLOWED_ORGS`.

Before this PR, only `{org}/.fullsend/` workflows could mint tokens, so the trust surface was narrower. Now that upstream workflows are accepted, a contributor who merges any workflow into the upstream repo (CI, test, docs build) could mint tokens.

Recommend: make `ALLOWED_WORKFLOW_FILES` fail closed. An unset or empty value should deny all requests. If the operator wants to allow any workflow file, they should explicitly set it to `*`. Also add a test for the unset/empty scenario — the existing `TestHandler_OIDCPrevalidation_WorkflowAllowlist` only tests the set case.

File: .github/workflows/reusable-retro.yml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        id: repo-parts
+        uses: fullsend-ai/fullsend/.github/actions/validate-enrollment@v0
+        with:
+          source_repo: ${{ inputs.source_repo }}
+
+      - name: Mint retro token
+        id: app-token
+        uses: fullsend-ai/fullsend/.github/actions/mint-token@v0
+        with:
+          role: fullsend
```
Comment (ralphbean) [inline .github/workflows/reusable-retro.yml]: **[important]** Role escalation: the old retro workflow used `role: retro` (grants `contents: read, pull_requests: read, issues: write`). This now uses `role: fullsend` which grants `contents: write, pull_requests: write, actions: write, workflows: write` — far beyond what a retrospective agent needs.

The `retro` role still exists in `rolePermissions`. Was this intentional? If the retro agent needs `.fullsend` repo write access, that rationale should be documented. Otherwise, revert to `role: retro`.

File: docs/ADRs/0035-layered-content-resolution.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+**C. Scaffold stops writing upstream defaults.** `WalkFullsendRepo` skips files
+in layered directories (`agents/`, `skills/`, `schemas/`, `harness/`,
+`policies/`, `scripts/`, `env/`) and upstream-only directories (`.github/actions/`,
+`.github/scripts/`). The installer writes only org-specific files and
+`customized/` gitkeeps.
+
+File categories after this change:
+
+- **Org-only** (~18 files): `env/`, `dispatch.yml`, thin callers, shim
```
Comment (ralphbean) [inline docs/ADRs/0035-layered-content-resolution.md]: **[important]** `env/` is listed here under "Org-only" but in the code it's a layered directory:
- `scaffold.go` `layeredDirs` includes `"env/"`
- `customized/env/.gitkeep` exists in the scaffold
- All 5 reusable workflows include `env` in `LAYERED_DIRS`
- `TestLayeredDirsNotInstalled` lists `"env/"` as a skipped prefix

`env/` should be moved to the "Upstream defaults" category (or "Org overrides" if that's more accurate). Also, the "Org overrides" bullet says "7 `.gitkeep` files" but the parenthetical only lists 6 subdirs — `env/` is the missing 7th.

File: internal/scaffold/scaffold_test.go [hunk 4]
```diff
@@ -233,7 +230,77 @@ func TestWalkFullsendRepo(t *testing.T) {
...
+}
+
+func TestCustomizedDirsInstalled(t *testing.T) {
+	expected := map[string]bool{
+		"customized/agents/.gitkeep":   false,
+		"customized/skills/.gitkeep":   false,
+		"customized/schemas/.gitkeep":  false,
+		"customized/harness/.gitkeep":  false,
+		"customized/policies/.gitkeep": false,
+		"customized/scripts/.gitkeep":  false,
```
Comment (ralphbean) [inline internal/scaffold/scaffold_test.go:266]: **[minor]** Missing `"customized/env/.gitkeep": false` — the scaffold has this file, `env/` is in `layeredDirs`, and the reusable workflows layer `env/`. The test should verify all 7 gitkeeps, not just 6. (The e2e test at `verifyInstalled` line 407 does check for it, so this is a unit-test gap rather than a coverage gap.)

File: .github/workflows/reusable-code.yml [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      - name: Checkout upstream defaults
+        uses: actions/checkout@v6
+        with:
+          repository: fullsend-ai/fullsend
+          ref: v0
+          path: .defaults
+          sparse-checkout: |
+            internal/scaffold/fullsend-repo/
+
+      - name: Prepare workspace (upstream defaults + org overrides)
```
Comment (ralphbean) [inline .github/workflows/reusable-code.yml:58]: **[moderate] (noted, deferred)** This "Prepare workspace" step is byte-for-byte identical across all 5 reusable workflows (`reusable-code`, `reusable-fix`, `reusable-review`, `reusable-triage`, `reusable-retro`), with a slightly different variant in `prioritize.yml`. If the layering logic or `LAYERED_DIRS` changes, all 6 must be updated in lockstep. Consider extracting to a composite action (`.github/actions/prepare-workspace/action.yml`) consistent with the existing `mint-token`, `setup-gcp`, and `validate-enrollment` actions.

File: internal/scaffold/scaffold.go [hunk 6]
```diff
@@ -47,22 +48,72 @@ func FileMode(path string) string {
 	return "100644"
 }
 
-// WalkFullsendRepo calls fn for each file in the fullsend-repo scaffold.
-// Paths passed to fn are relative to the fullsend-repo root.
+// layeredDirs contain upstream defaults provided at runtime via reusable
+// workflow workspace preparation. The scaffold does not install these —
+// orgs add overrides in customized/<dir>/ instead. See ADR 0035.
+var layeredDirs = []string{
```
Comment (ralphbean) [inline internal/scaffold/scaffold.go:54]: **[moderate] (noted, deferred)** The `layeredDirs` list here must stay in sync with `LAYERED_DIRS` in all 5 reusable workflows. Currently both list `agents skills schemas harness policies scripts env`, but there's no automated test to catch drift. Consider adding a test that parses `LAYERED_DIRS` from a reusable workflow YAML and compares it against this Go slice.

File: internal/scaffold/fullsend-repo/templates/shim-workflow-call.yaml [hunk 7]
```diff
@@ -33,175 +34,19 @@ on:
     types: [submitted]
 
 jobs:
-  dispatch-triage:
+  dispatch:
```
Comment (ralphbean) [inline internal/scaffold/fullsend-repo/templates/shim-workflow-call.yaml:37]: **[moderate] (noted, deferred)** The old shim's `post-run-link` job ("fullsend X is working on this — view logs") is removed. This is user-visible — users will no longer get feedback that an agent was dispatched. Plan is to restore this from inside `fullsend run` so it uses the correct app identity token (not `github-actions[bot]`) and can provide a better link.

File: internal/scaffold/fullsend-repo/.github/workflows/dispatch.yml [hunk 8]
```diff
@@ -24,7 +23,170 @@ jobs:
       contents: read
       id-token: write
     steps:
+      - name: Determine stage
```
Comment (ralphbean) [inline internal/scaffold/fullsend-repo/.github/workflows/dispatch.yml:26]: **[moderate] (noted, deferred)** The routing logic here (~100 lines of bash: command parsing, label matching, author association, bot filtering, fork detection, auto-triage) has no automated test coverage. Any regression silently breaks all dispatch. The old shim used GHA `if:` expressions which were at least visible in the workflow graph. Consider adding a shell-script test that exercises this routing with various event/input combinations.

**Review by ralphbean — CHANGES_REQUESTED** [review]
## Review Summary

This is a well-motivated architectural improvement implementing three ADRs (31, 34, 35) that dramatically reduces per-org maintenance burden. The security posture is strong — expression injection avoidance, explicit secret passing, fork PR blocking, and OIDC validation are all correctly implemented.

Since our last review pass (against `b1dbc1d`), three issues have been fixed:
- ✅ SPEC.md for ADR 0012 removed (was normatively incorrect)
- ✅ Fork PR `|| true` checks now fail closed with explicit empty-string guards
- ✅ e2e `verifyInstalled` updated correctly for ADR 35

**4 items require changes:**
1. Mint token validation fails open when `ALLOWED_WORKFLOW_FILES` is unset — should fail closed
2. Retro agent role escalated from `retro` to `fullsend` without justification
3. ADR 0035 miscategorizes `env/` as org-only when it's actually a layered directory
4. Test missing `customized/env/.gitkeep` assertion

**4 items noted for follow-up:**
- Duplicated "Prepare workspace" step across 5 reusable workflows (extract to composite action)
- No test verifying LAYERED_DIRS consistency between Go and YAML
- `post-run-link` feedback removed (restore inside `fullsend run` with correct app identity)
- No automated tests for dispatch routing logic (~100 lines of bash)

File: internal/mint/main.go (hunk 1 above)
Comment (waynesun09) [reply internal/mint/main.go:577]: Fixed in 1cabedaf. `ALLOWED_WORKFLOW_FILES` now fails closed — unset/empty denies all requests. Added `*` wildcard for explicit allow-all. New test `TestHandler_OIDCPrevalidation_WorkflowAllowlistUnset` covers the empty scenario.

File: .github/workflows/reusable-retro.yml (hunk 2 above)
Comment (waynesun09) [reply .github/workflows/reusable-retro.yml]: Fixed in 1cabedaf. Reverted to `role: retro` — the retro agent only needs `contents: read, pull_requests: read, issues: write`, which the `retro` role already provides.

File: docs/ADRs/0035-layered-content-resolution.md (hunk 3 above)
Comment (waynesun09) [reply docs/ADRs/0035-layered-content-resolution.md]: Fixed in 1cabedaf. Moved `env/` from Org-only to Upstream defaults. Updated Org overrides parenthetical to list all 7 subdirs: `customized/{agents,skills,schemas,harness,policies,scripts,env}/`.

File: internal/scaffold/scaffold_test.go (hunk 4 above)
Comment (waynesun09) [reply internal/scaffold/scaffold_test.go:266]: Fixed in 1cabedaf. Added `"customized/env/.gitkeep": false` to `TestCustomizedDirsInstalled` — now asserts all 7 gitkeeps.

File: internal/scaffold/fullsend-repo/.github/workflows/dispatch.yml [hunk 9]
```diff
@@ -56,25 +218,25 @@ jobs:
           echo "token=$TOKEN" >> "$GITHUB_OUTPUT"
 
       - name: Checkout repository
+        if: steps.route.outputs.stage != ''
         uses: actions/checkout@v6
```
Comment (ralphbean) [inline internal/scaffold/fullsend-repo/.github/workflows/dispatch.yml:222]: **[moderate] (noted, deferred)** This checkout uses the minted OIDC token but doesn't set `persist-credentials: false`. The token stays in the git credential helper for the rest of the job. While dispatch.yml only runs trusted shell scripts (not LLM-driven), the reusable workflows correctly set `persist-credentials: false` on all their target-repo checkouts. For consistency and defense-in-depth, consider adding `persist-credentials: false` here too.

File: e2e/admin/admin_test.go [hunk 10]
```diff
@@ -385,38 +385,26 @@ func verifyInstalled(t *testing.T, env *e2eEnv, orgCfg *config.OrgConfig, enable
...
-		"harness/fix.yaml",
-		"policies/fix.yaml",
-		"env/fix-agent.env",
-		"schemas/fix-result.schema.json",
-		"scripts/pre-fix.sh",
-		"scripts/post-fix.sh",
-		"scripts/process-fix-result.py",
+		".github/workflows/dispatch.yml",
+		".github/workflows/repo-maintenance.yml",
+		".github/workflows/prioritize.yml",
```
Comment (ralphbean) [inline e2e/admin/admin_test.go:399]: **[moderate] (noted, deferred)** This list is missing `.github/workflows/retro.yml` and `AGENTS.md` — both are in the scaffold, not in any layered or upstream-only directory, so `WalkFullsendRepo` will install them. Without assertions here, regressions on those files won't be caught by e2e tests.

**Review by ralphbean — APPROVED** [review]
## Review Summary

All 4 required changes from the previous review have been addressed in `1cabedaf`:
- ✅ `ALLOWED_WORKFLOW_FILES` now fails closed (unset/empty denies all)
- ✅ Retro role reverted from `fullsend` to `retro`
- ✅ ADR 0035 `env/` moved from Org-only to Upstream defaults
- ✅ `customized/env/.gitkeep` assertion added to scaffold test

The security posture is strong: expression injection avoidance via `env:` blocks, explicit secret passing (not `secrets: inherit`), fork PR fail-closed detection, `persist-credentials: false` on all target-repo checkouts in reusable workflows, and OIDC `job_workflow_ref` validation with a fail-closed allowlist.

**2 items noted for follow-up** (not blocking):
1. `dispatch.yml` checkout missing `persist-credentials: false` on minted token checkout
2. e2e `verifyInstalled` missing `retro.yml` and `AGENTS.md` assertions
